### PR TITLE
[AC-1486] Feature: SM Billing Round 1

### PR DESCRIPTION
**Pull request by vvolkgang**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Long-lived feature branch to support billing for Secrets Manager.

Requires server PR: https://github.com/bitwarden/server/pull/3073

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: libs/common/src/billing/models/response/plan.response.ts [hunk 1]
```diff
@@ -48,6 +49,15 @@ export class PlanResponse extends BaseResponse {
...
   premiumAccessOptionPrice: number;
 
+  // SM only
+  additionalPricePerServiceAccount: number;
+  baseServiceAccount: number;
+  maxServiceAccount: number;
+  hasAdditionalServiceAccountOption: boolean;
+  maxProjects: number;
+  maxAdditionalServiceAccounts: number;
+  stripeServiceAccountPlanId: string;
```
Comment (Hinton) [inline libs/common/src/billing/models/response/plan.response.ts:59]: I would kind of have expected there to be two plan models, one for PM and another for SM. That both extends a common model?

File: libs/common/src/admin-console/models/response/organization.response.ts [hunk 2]
```diff
@@ -27,6 +28,11 @@ export class OrganizationResponse extends BaseResponse {
   useResetPassword: boolean;
   useSecretsManager: boolean;
   hasPublicAndPrivateKeys: boolean;
+  usePasswordManager: boolean;
```
Comment (Hinton) [inline libs/common/src/admin-console/models/response/organization.response.ts:31]: Do we need this if we have the `plan` from which we can deduct that they use pm? (Same for `useSecretsManager`)

**Review by Hinton — APPROVED** [review]
Platform changes looks good. Some confusion around how the billing portions are implemented. In some places we reference a SM plan, but not in create or update? And plans seems to be a union between PM and SM.

File: libs/common/src/billing/models/response/subscription.response.ts [hunk 3]
```diff
@@ -62,6 +63,8 @@ export class BillingSubscriptionItemResponse extends BaseResponse {
   quantity: number;
   interval: string;
   sponsoredSubscriptionItem: boolean;
+  addonSubscriptionItem: boolean;
+  bitwardenProduct: BitwardenProductType;
```
[the remark this answers is not in the document]
Comment (eliykat) [reply libs/common/src/billing/models/response/subscription.response.ts:67]: We already have a `ProductType` enum, which is basically the same as the `PlanType` but without any reference to monthly/annually billing cycle or the year:

```ts
export enum ProductType {
  Free = 0,
  Families = 1,
  Teams = 2,
  Enterprise = 3,
}
```

We might be able to reduce the duplication between `PlanType`/`ProductType` (which in turn would reduce this confusion) when we refactor plans.

File: libs/common/src/admin-console/models/response/organization.response.ts (hunk 2 above)
Comment (eliykat) [reply libs/common/src/admin-console/models/response/organization.response.ts:31]: The `plan` object shows what the plan supports (in general, e.g. enterprise plans support different features to teams plans), not what this particular organization is using.

File: libs/common/src/billing/models/response/plan.response.ts (hunk 1 above)
Comment (eliykat) [reply libs/common/src/billing/models/response/plan.response.ts:59]: Yes, the plan models are the biggest thing we need to refactor post GA. I actually think there should be 1 plan object per subscription level (e.g. teams/enterprise/families/free) showing what features and products that subscription level supports. Given that an org can only ever have 1 subscription level. Happy to discuss further if you like, but in any case this is top of my list for followup.

**Review by eliykat — COMMENTED** [review]
Approved in principle, but I've done some of this work so I will ask for a final review and merge from either @cyprain-okeke or @r-tome.

✅  each commit is approved in a separate PR or is otherwise approved now
✅  list of change files does not show any unexpected changes

**Review by r-tome — APPROVED** [review]

**Review by cyprain-okeke — APPROVED** [review]
Looks Great

**Review by eliykat — APPROVED** [review]

File: libs/common/src/billing/enums/bitwarden-product-type.enum.ts [hunk 4]
```diff
@@ -0,0 +1,4 @@
+export enum BitwardenProductType {
```
Comment (MGibson1) [inline libs/common/src/billing/enums/bitwarden-product-type.enum.ts:1]: `ProductType` is an existing enum that defines free, teams, enterprise, etc.

Can we rename this to `BitwardenProduct`? `Type` really isn't adding anything to this name anyway.

File: libs/common/src/billing/models/request/organization-subscription-update.request.ts
Comment (MGibson1) [inline libs/common/src/billing/models/request/organization-subscription-update.request.ts:1]: yay for docs! boo that we lose our inline properties

**Review by MGibson1 — APPROVED** [review]

**Pull request merged**